android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt

```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-        Notification.Builder(context, CHANNEL_ID)
-    else
-        Notification.Builder(context)
-
+private fun createViewPendingIntent(fcmMessage: MessageFcmMessage, context: Context): PendingIntent {
     val uri = Uri.fromParts("zulip", "msgid:${fcmMessage.zulipMessageId}", "")
     val viewIntent = Intent(Intent.ACTION_VIEW, uri, context, NotificationIntentService::class.java)
     viewIntent.putExtra(EXTRA_NOTIFICATION_DATA, fcmMessage.dataForOpen())
-    val viewPendingIntent = PendingIntent.getService(context, 0, viewIntent, 0)
-    builder.setContentIntent(viewPendingIntent)
-    builder.setAutoCancel(true)
+    return PendingIntent.getService(context, 0, viewIntent, 0)
+}
 
-    val totalMessagesCount = extractTotalMessagesCount(conversations)
+private fun createDismissAction(context: Context): NotificationCompat.Action {
+    val dismissIntent = Intent(context, NotificationIntentService::class.java)
+    dismissIntent.action = ACTION_CLEAR
+    val piDismiss = PendingIntent.getService(context, 0, dismissIntent, 0)
+    return NotificationCompat.Action(
+        android.R.drawable.ic_menu_close_clear_cancel,
+        "Clear",
+        piDismiss
+    )
+}
 
-    if (BuildConfig.DEBUG) {
-        builder.setSmallIcon(R.mipmap.ic_launcher)
-    } else {
-        builder.setSmallIcon(R.drawable.zulip_notification)
+/**
+ * @param context
+ * @param conversationKey Unique Key identifying a conversation, the current
+ * implementation assumes that each notification corresponds to 1 and only 1
+ * conversation.
+ *
+ * A conversation can be any of: Topic in Stream, GroupPM or PM for a
+ * specific user in a specific realm.
+ */
+private fun getActiveNotification(context: Context, conversationKey: String):
+    Pair<Int?, Notification?> {
+    // activeNotifications are not available in NotificationCompatManager
+    // Hence we have to use instance of NotificationManager.
+    val notificationManager =
+        context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
+
+    val activeStatusBarNotifications = notificationManager?.activeNotifications
+    if (activeStatusBarNotifications != null) {
+        for (statusBarNotification in activeStatusBarNotifications) {
+            val notification = statusBarNotification.notification
+            if (notification.extras.getString("conversationKey") == conversationKey) {
+                return Pair(statusBarNotification.id, notification)
+            }
+        }
     }
+    return Pair(null, null)
+}
 
-    // This should agree with `BRAND_COLOR` in the JS code.
-    builder.setColor(Color.rgb(100, 146, 254))
-
-    val nameList = extractNames(conversations)
-
-    if (conversations.size == 1 && nameList.size == 1) {
-        //Only one 1 notification therefore no using of big view styles
-        if (totalMessagesCount > 1) {
-            builder.setContentTitle("${fcmMessage.sender.fullName} ($totalMessagesCount)")
+private fun createSummaryNotification(
```

Comment:
> How did you find that it's required by Android even on newer versions?

I tested a build without summary Notification, specifically without the notify call to summaryNotification, observation is that notification do occur but they don't group together, even in android 7+. It seems that this notify call triggers the actual grouping.

Also while they don't appear completely in Android 7+, (since they get overridden by the summary notification built by the Android OS) they still have some visual influence on how the summary notification looks, for instance, setting
`summaryText` shows as the [header text](https://material.io/design/platform-guidance/android-notifications.html#anatomy-of-a-notification) in summary notification.

Reply:
> I tested a build without summary Notification, specifically without the notify call to summaryNotification, observation is that notification do occur but they don't group together, even in android 7+. It seems that this notify call triggers the actual grouping.

Hmm fascinating -- thanks for determining that. Seems like a bug [in the docs](https://developer.android.com/training/notify-user/group) that they say it's optional, then.

I guess in the example code says:
```kt
val summaryNotification = NotificationCompat.Builder(this@MainActivity, CHANNEL_ID)
        .setContentTitle(emailObject.getSummary())
        //set content text to support devices running API level < 24
        .setContentText("Two new messages")
        // […]
```
which makes it sound more like the actual situation you discovered, that the summary notification itself is required but the text is only for older devices.

We did [find when I looked recently](https://chat.zulip.org/#narrow/stream/48-mobile/topic/platform.20versions/near/1254509) that indeed Android 6 and older is down to a small enough sliver of our userbase that it makes sense to drop support for it as soon as there's something where supporting it would make things more complicated for us. So I think we won't go and add a summary text, and instead will just drop support and increase the minimum to Android 7.